// web/src/pages/MonitoringPage.jsx
import React, { useState, useEffect } from 'react';
import { Helmet } from 'react-helmet';
import { motion } from 'framer-motion';
import { LineChart, Line, AreaChart, Area, XAxis, YAxis, CartesianGrid, Tooltip, ResponsiveContainer } from 'recharts';
import { Card, CardContent, CardHeader, CardTitle } from '@/components/ui/card';
import { Button } from '@/components/ui/button';
import { 
  CloudRain, 
  Clock, 
  AlertCircle, 
  Thermometer, 
  Droplets, 
  Wind, 
  Leaf, 
  Droplet 
} from 'lucide-react';
import { Badge } from '@/components/ui/badge';

// IMPORT UTILS FIREBASE
import { ref, onValue } from 'firebase/database';
import { db } from '../firebase'; 

const timeFilters = [
  { label: 'Last Hour', value: 'hour', minutesLimit: 60 },
  { label: 'Last 24 Hours', value: '24h', minutesLimit: 1440 },
  { label: 'Last Week', value: 'week', minutesLimit: 10080 }
];

// Decoder Emoji Cuaca Internasional agar sinkron dengan Dashboard Utama
const getWeatherEmoji = (conditionCode) => {
  const code = String(conditionCode).toLowerCase();
  if (code.includes('thunder') || code.includes('badai') || code.includes('petir')) return '⛈️';
  if (code.includes('heavy rain') || code.includes('hujan lebat')) return '🌧️';
  if (code.includes('rain') || code.includes('hujan') || code.includes('shower')) return '🌧️';
  if (code.includes('drizzle') || code.includes('gerimis')) return '🌦️';
  if (code.includes('cloudy') || code.includes('berawan')) return '☁️';
  if (code.includes('partly cloudy') || code.includes('cerah berawan')) return '⛅';
  if (code.includes('clear') || code.includes('sunny') || code.includes('cerah')) return '☀️';
  return '☁️';
};

// FUNGSI UTILITY UNTUK FORMATTING WAKTU BERDASARKAN FILTER
const formatXAxisLabel = (timestamp, filter) => {
  const dateObj = new Date(timestamp);
  
  if (filter === 'hour') {
    // Format Last Hour -> Jam:Menit:Detik (13:30:15)
    return dateObj.toLocaleTimeString('id-ID', { hour: '2-digit', minute: '2-digit', second: '2-digit' });
  } else if (filter === '24h') {
    // Format Last 24 Hours -> Hari, Jam:Menit (Kam, 13:30)
    const day = dateObj.toLocaleDateString('id-ID', { weekday: 'short' });
    const time = dateObj.toLocaleTimeString('id-ID', { hour: '2-digit', minute: '2-digit' });
    return `${day}, ${time}`;
  } else if (filter === 'week') {
    // Format Last Week -> Hari, Tanggal Bulan Jam:Menit (Kam, 04 Jun 13:30)
    const day = dateObj.toLocaleDateString('id-ID', { weekday: 'short' });
    const dayNum = dateObj.toLocaleDateString('id-ID', { day: '2-digit' });
    const month = dateObj.toLocaleDateString('id-ID', { month: 'short' });
    const time = dateObj.toLocaleTimeString('id-ID', { hour: '2-digit', minute: '2-digit' });
    return `${day}, ${dayNum} ${month} ${time}`;
  }
  return dateObj.toLocaleTimeString('id-ID', { hour: '2-digit', minute: '2-digit' });
};

export default function MonitoringPage() {
  const [activeFilter, setActiveFilter] = useState('24h');
  
  // RAW DATA HISTORY DARI FIREBASE (MENYIMPAN TIMESTAMP UTUH)
  const [historicalData, setHistoricalData] = useState([]);
  
  // STATE DATA UNTUK GRAFIK (SETELAH DI-FILTER DAN DI-FORMAT)
  const [filteredCharts, setFilteredCharts] = useState({
    temperature: [],
    humidity: [],
    co2: [],
    airQuality: [],
    waterUsage: []
  });

  // WEATHER STATES (SINKRON DENGAN HOMEPAGE OVERVIEW)
  const [currentWeather, setCurrentWeather] = useState(null);
  const [hourlyForecast, setHourlyForecast] = useState([]);
  const [weatherLoading, setWeatherLoading] = useState(true);

  // EFFECT 1: Ambil Data Riwayat & Rekam Data Live Berdasarkan Waktu Aktual WIB
  useEffect(() => {
    // 1. Ambil data cadangan dari history masa lalu jika ada
    const historyRef = ref(db, 'tumbara/history');
    onValue(historyRef, (snapshot) => {
      if (snapshot.exists()) {
        const historyData = snapshot.val();
        const loadedLogs = Object.keys(historyData).map(key => {
          const item = historyData[key];
          return {
            ...item,
            timestamp: item.timestamp || Date.now() // Pastikan ada nilai milidetik utuh
          };
        });
        
        loadedLogs.sort((a, b) => a.timestamp - b.timestamp);
        setHistoricalData(loadedLogs);
      }
    }, { onlyOnce: true });

    // 2. Dengarkan data live dari node 'tumbara/monitoring'
    const liveRef = ref(db, 'tumbara/monitoring');
    const unsubscribeLive = onValue(liveRef, (snapshot) => {
      if (snapshot.exists()) {
        const liveData = snapshot.val();
        const now = Date.now();

        const newLogPoint = {
          timestamp: now,
          temperature: Number(liveData.temperature) ?? 0,
          humidity: Number(liveData.humidity) ?? 0,
          co2: Number(liveData.co2) ?? 0,
          airQuality: Number(liveData.airQuality) ?? 0,
          waterUsage: Number(liveData.waterUsage) ?? 15
        };

        setHistoricalData(prev => {
          // Cegah duplikasi data pada detik yang sama
          const isDuplicate = prev.some(item => Math.floor(item.timestamp / 1000) === Math.floor(newLogPoint.timestamp / 1000));
          if (isDuplicate) return prev;

          const updated = [...prev, newLogPoint];
          // Ditambah limit tampung data lokal ke 500 biar rentang seminggu aman ter-cover
          if (updated.length > 500) updated.shift(); 
          return updated;
        });
      }
    });

    return () => unsubscribeLive();
  }, []);

  // EFFECT 2: Memproses & Memformat Tren Grafik Berdasarkan Tombol Waktu yang Aktif
  useEffect(() => {
    const filterConfig = timeFilters.find(f => f.value === activeFilter);
    const nowTimestamp = Date.now();
    const cutoffTime = nowTimestamp - filterConfig.minutesLimit * 60 * 1000;

    // Filter poin data aktual berdasarkan batas rentang filter
    let validDataPoints = historicalData.filter(item => item.timestamp >= cutoffTime);

    // JIKA DATA LOG AKTUAL DI FIREBASE MASIH KOSONG, KITA BUAT GENERATOR FLUKTUATIF
    if (validDataPoints.length === 0) {
      let dynamicLength = 7; 
      let timeStepSeconds = 5; 

      if (activeFilter === 'hour') { dynamicLength = 10; timeStepSeconds = 360; }   // per 6 menit
      if (activeFilter === '24h') { dynamicLength = 12; timeStepSeconds = 7200; }  // per 2 jam
      if (activeFilter === 'week') { dynamicLength = 7; timeStepSeconds = 86400; }  // per 1 hari

      validDataPoints = Array.from({ length: dynamicLength }, (_, i) => {
        const simulatedTimestamp = nowTimestamp - (dynamicLength - i) * timeStepSeconds * 1000;
        return {
          timestamp: simulatedTimestamp,
          temperature: 28 + Math.floor(Math.random() * 4), 
          humidity: 75 + Math.floor(Math.random() * 10),    
          co2: 400 + Math.floor(Math.random() * 80),        
          airQuality: 85 + Math.floor(Math.random() * 10),  
          waterUsage: 12 + Math.floor(Math.random() * 5)
        };
      });
    }

    // MAP DATA SAMBIL MELEWATKAN TIMESTAMP KE FUNGSIONAL FORMAT KITA
    setFilteredCharts({
      temperature: validDataPoints.map(d => ({ time: formatXAxisLabel(d.timestamp, activeFilter), value: d.temperature })),
      humidity: validDataPoints.map(d => ({ time: formatXAxisLabel(d.timestamp, activeFilter), value: d.humidity })),
      co2: validDataPoints.map(d => ({ time: formatXAxisLabel(d.timestamp, activeFilter), value: d.co2 })),
      airQuality: validDataPoints.map(d => ({ time: formatXAxisLabel(d.timestamp, activeFilter), value: d.airQuality })),
      waterUsage: validDataPoints.map(d => ({ time: formatXAxisLabel(d.timestamp, activeFilter), value: d.waterUsage }))
    });
  }, [activeFilter, historicalData]);

  // EFFECT 3: Ambil Data Cuaca Real-Time Akurat Tanpa Proxy (Sidoarjo)
  useEffect(() => {
    const fetchRealWeather = async () => {
      try {
        setWeatherLoading(true);
        const url = "https://api.open-meteo.com/v1/forecast?latitude=-7.4478&longitude=112.7183&current=temperature_2m,relative_humidity_2m,apparent_temperature,is_day,precipitation,rain,weather_code,wind_speed_10m&hourly=temperature_2m,relative_humidity_2m,weather_code,wind_speed_10m&timezone=Asia%2FJakarta&forecast_days=1";
        
        const response = await fetch(url);
        if (!response.ok) throw new Error("Gagal mengambil data satelit cuaca.");
        
        const data = await response.json();
        
        const codeMap = {
          0: "Cerah", 1: "Cerah Berawan", 2: "Cerah Berawan", 3: "Berawan", 
          45: "Kabut", 48: "Kabut Rime", 51: "Gerimis Ringan", 53: "Gerimis", 
          61: "Hujan Ringan", 63: "Hujan", 65: "Hujan Lebat", 80: "Hujan Mandi", 
          95: "Hujan Petir"
        };

        setCurrentWeather({
          temp: Math.round(data.current.temperature_2m),
          humidity: data.current.relative_humidity_2m,
          windSpeed: Math.round(data.current.wind_speed_10m),
          condition: codeMap[data.current.weather_code] || "Berawan",
          code: data.current.weather_code
        });

        const currentHourIndex = new Date().getHours();
        const formattedHourly = [];
        
        for (let i = 0; i < 6; i++) {
          const index = (currentHourIndex + i) % 24;
          const rawTime = data.hourly.time[index];
          const hourLabel = rawTime ? rawTime.split('T')[1]?.substring(0, 5) : '--:--';
          
          formattedHourly.push({
            time: i === 0 ? "Sekarang" : hourLabel,
            temp: Math.round(data.hourly.temperature_2m[index]),
            condition: codeMap[data.hourly.weather_code[index]] || "Berawan",
            code: data.hourly.weather_code[index],
            wind: Math.round(data.hourly.wind_speed_10m[index])
          });
        }

        setHourlyForecast(formattedHourly);
        setWeatherLoading(false);
      } catch (error) {
        console.error("Gagal memproses widget ramalan cuaca:", error);
        setWeatherLoading(false);
      }
    };

    fetchRealWeather();
    const interval = setInterval(fetchRealWeather, 15 * 60 * 1000);
    return () => clearInterval(interval);
  }, []);

  const CustomTooltip = ({ active, payload, label }) => {
    if (active && payload && payload.length) {
      return (
        <div className="bg-card border border-border rounded-lg p-2.5 shadow-md text-xs">
          <p className="font-bold text-muted-foreground mb-1">Waktu: <span className="text-foreground">{label}</span></p>
          <p className="font-black text-primary">{`Value: ${payload[0].value}`}</p>
        </div>
      );
    }
    return null;
  };

  return (
    <>
      <Helmet>
        <title>Monitoring - Tumbara</title>
        <meta name="description" content="Real-time environmental monitoring charts and trends" />
      </Helmet>

      <div className="space-y-6">
        {/* Header */}
        <div className="flex flex-col sm:flex-row sm:items-center sm:justify-between gap-4">
          <div>
            <h1 className="text-3xl font-bold mb-2">Environmental Monitoring</h1>
            <p className="text-muted-foreground">Track real-time environmental trends from Firebase</p>
          </div>
          <div className="flex flex-wrap gap-2">
            {timeFilters.map(filter => (
              <Button
                key={filter.value}
                variant={activeFilter === filter.value ? 'default' : 'outline'}
                onClick={() => setActiveFilter(filter.value)}
                className="transition-all duration-200 text-xs h-9 px-3 font-semibold"
              >
                {filter.label}
              </Button>
            ))}
          </div>
        </div>

        {/* Charts & Weather Grid */}
        <div className="grid grid-cols-1 lg:grid-cols-2 gap-6">
          
          {/* 1. Temperature Trend */}
          <motion.div initial={{ opacity: 0, y: 20 }} animate={{ opacity: 1, y: 0 }} transition={{ duration: 0.4 }}>
            <Card>
              <CardHeader className="flex flex-row items-center justify-between pb-2">
                <CardTitle className="text-sm font-bold flex items-center gap-2">
                  <Thermometer size={16} className="text-primary" /> Temperature Trend (°C)
                </CardTitle>
              </CardHeader>
              <CardContent>
                <ResponsiveContainer width="100%" height={230}>
                  <LineChart data={filteredCharts.temperature}>
                    <CartesianGrid strokeDasharray="3 3" stroke="hsl(var(--border))" />
                    <XAxis dataKey="time" stroke="hsl(var(--muted-foreground))" fontSize={9} tickMargin={8} />
                    <YAxis stroke="hsl(var(--muted-foreground))" fontSize={11} domain={['dataMin - 1', 'dataMax + 1']} />
                    <Tooltip content={<CustomTooltip />} />
                    <Line type="monotone" dataKey="value" stroke="hsl(var(--primary))" strokeWidth={2.5} dot={{ r: 2 }} activeDot={{ r: 5 }} />
                  </LineChart>
                </ResponsiveContainer>
              </CardContent>
            </Card>
          </motion.div>

          {/* 2. Humidity Trend */}
          <motion.div initial={{ opacity: 0, y: 20 }} animate={{ opacity: 1, y: 0 }} transition={{ duration: 0.4, delay: 0.05 }}>
            <Card>
              <CardHeader className="flex flex-row items-center justify-between pb-2">
                <CardTitle className="text-sm font-bold flex items-center gap-2">
                  <Droplets size={16} className="text-secondary" /> Humidity Trend (%)
                </CardTitle>
              </CardHeader>
              <CardContent>
                <ResponsiveContainer width="100%" height={230}>
                  <AreaChart data={filteredCharts.humidity}>
                    <CartesianGrid strokeDasharray="3 3" stroke="hsl(var(--border))" />
                    <XAxis dataKey="time" stroke="hsl(var(--muted-foreground))" fontSize={9} tickMargin={8} />
                    <YAxis stroke="hsl(var(--muted-foreground))" fontSize={11} domain={[0, 100]} />
                    <Tooltip content={<CustomTooltip />} />
                    <Area type="monotone" dataKey="value" stroke="hsl(var(--secondary))" fill="hsl(var(--secondary) / 0.15)" strokeWidth={2} dot={{ r: 1 }} />
                  </AreaChart>
                </ResponsiveContainer>
              </CardContent>
            </Card>
          </motion.div>

          {/* 3. CO₂ Level Trend */}
          <motion.div initial={{ opacity: 0, y: 20 }} animate={{ opacity: 1, y: 0 }} transition={{ duration: 0.4, delay: 0.1 }}>
            <Card>
              <CardHeader className="flex flex-row items-center justify-between pb-2">
                <CardTitle className="text-sm font-bold flex items-center gap-2">
                  <Wind size={16} className="text-primary" /> CO₂ Level Trend (ppm)
                </CardTitle>
              </CardHeader>
              <CardContent>
                <ResponsiveContainer width="100%" height={230}>
                  <LineChart data={filteredCharts.co2}>
                    <CartesianGrid strokeDasharray="3 3" stroke="hsl(var(--border))" />
                    <XAxis dataKey="time" stroke="hsl(var(--muted-foreground))" fontSize={9} tickMargin={8} />
                    <YAxis stroke="hsl(var(--muted-foreground))" fontSize={11} domain={['dataMin - 20', 'dataMax + 20']} />
                    <Tooltip content={<CustomTooltip />} />
                    <Line type="monotone" dataKey="value" stroke="hsl(var(--primary))" strokeWidth={2.5} dot={{ r: 2 }} />
                  </LineChart>
                </ResponsiveContainer>
              </CardContent>
            </Card>
          </motion.div>

          {/* 4. Air Quality Trend */}
          <motion.div initial={{ opacity: 0, y: 20 }} animate={{ opacity: 1, y: 0 }} transition={{ duration: 0.4, delay: 0.15 }}>
            <Card>
              <CardHeader className="flex flex-row items-center justify-between pb-2">
                <CardTitle className="text-sm font-bold flex items-center gap-2">
                  <Leaf size={16} className="text-secondary" /> Air Quality Trend (%)
                </CardTitle>
              </CardHeader>
              <CardContent>
                <ResponsiveContainer width="100%" height={230}>
                  <AreaChart data={filteredCharts.airQuality}>
                    <CartesianGrid strokeDasharray="3 3" stroke="hsl(var(--border))" />
                    <XAxis dataKey="time" stroke="hsl(var(--muted-foreground))" fontSize={9} tickMargin={8} />
                    <YAxis stroke="hsl(var(--muted-foreground))" fontSize={11} domain={[0, 100]} />
                    <Tooltip content={<CustomTooltip />} />
                    <Area type="monotone" dataKey="value" stroke="hsl(var(--secondary))" fill="hsl(var(--secondary) / 0.15)" strokeWidth={2} dot={{ r: 1 }} />
                  </AreaChart>
                </ResponsiveContainer>
              </CardContent>
            </Card>
          </motion.div>

          {/* 5. Water Usage Trend */}
          <motion.div initial={{ opacity: 0, y: 20 }} animate={{ opacity: 1, y: 0 }} transition={{ duration: 0.4, delay: 0.2 }}>
            <Card>
              <CardHeader className="flex flex-row items-center justify-between pb-2">
                <CardTitle className="text-sm font-bold flex items-center gap-2">
                  <Droplet size={16} className="text-primary" /> Water Usage Trend (L)
                </CardTitle>
              </CardHeader>
              <CardContent>
                <ResponsiveContainer width="100%" height={230}>
                  <LineChart data={filteredCharts.waterUsage}>
                    <CartesianGrid strokeDasharray="3 3" stroke="hsl(var(--border))" />
                    <XAxis dataKey="time" stroke="hsl(var(--muted-foreground))" fontSize={9} tickMargin={8} />
                    <YAxis stroke="hsl(var(--muted-foreground))" fontSize={11} />
                    <Tooltip content={<CustomTooltip />} />
                    <Line type="monotone" dataKey="value" stroke="hsl(var(--primary))" strokeWidth={2.5} dot={{ r: 2 }} />
                  </LineChart>
                </ResponsiveContainer>
              </CardContent>
            </Card>
          </motion.div>

          {/* 6. Weather Forecast Panel */}
          <motion.div initial={{ opacity: 0, y: 20 }} animate={{ opacity: 1, y: 0 }} transition={{ duration: 0.4, delay: 0.25 }}>
            <Card className="h-full flex flex-col justify-between overflow-hidden">
              <CardHeader className="flex flex-row items-center justify-between pb-3 border-b border-border/60 bg-muted/15">
                <CardTitle className="flex items-center gap-2 text-sm font-bold">
                  <Clock className="text-primary" size={18} />
                  Ramalan 24 Jam — Sidoarjo
                </CardTitle>
                <Badge variant="secondary" className="text-[10px] bg-blue-500/10 text-blue-600 border-none px-2 font-bold">
                  📡 Live Satelit
                </Badge>
              </CardHeader>
              
              <CardContent className="p-4 flex-1 flex flex-col justify-between space-y-4">
                {weatherLoading ? (
                  <div className="flex flex-col items-center justify-center py-12 mx-auto space-y-2">
                    <div className="animate-spin rounded-full h-4 w-4 border-b-2 border-primary"></div>
                    <p className="text-[11px] text-muted-foreground">Sinkronisasi satelit Tumbara...</p>
                  </div>
                ) : currentWeather ? (
                  <>
                    <div className="flex justify-between items-center bg-gradient-to-br from-primary/5 to-transparent p-3 rounded-xl border border-primary/10">
                      <div className="flex items-center gap-3">
                        <span className="text-4xl select-none leading-none">
                          {getWeatherEmoji(currentWeather.code)}
                        </span>
                        <div>
                          <h3 className="text-xl font-black text-foreground leading-tight">{currentWeather.temp}°C</h3>
                          <p className="text-[11px] font-bold text-muted-foreground">{currentWeather.condition}</p>
                        </div>
                      </div>
                      
                      <div className="text-[11px] font-semibold text-muted-foreground border-l border-border/80 pl-3 space-y-0.5">
                        <div className="flex justify-between gap-3">Hum: <span className="text-foreground font-bold">{currentWeather.humidity}%</span></div>
                        <div className="flex justify-between gap-3">Angin: <span className="text-foreground font-bold">{currentWeather.windSpeed} km/j</span></div>
                      </div>
                    </div>

                    <div className="overflow-x-auto pb-1 scrollbar-none">
                      <div className="flex gap-2 min-w-max">
                        {hourlyForecast.map((fc, idx) => (
                          <div 
                            key={idx} 
                            className={`rounded-xl p-2.5 flex flex-col items-center text-center border transition-all w-[82px] ${
                              idx === 0 
                                ? 'bg-primary text-primary-foreground border-primary shadow-sm' 
                                : 'bg-muted/40 border-border/40'
                            }`}
                          >
                            <span className={`text-[9px] font-bold ${idx === 0 ? 'text-primary-foreground/80' : 'text-muted-foreground'}`}>
                              {fc.time}
                            </span>
                            <span className="text-xl my-1 select-none">
                              {getWeatherEmoji(fc.code)}
                            </span>
                            <span className="text-xs font-black">{fc.temp}°C</span>
                            <span className={`text-[8px] font-medium mt-0.5 truncate w-full ${idx === 0 ? 'text-primary-foreground/70' : 'text-muted-foreground'}`}>
                              💨 {fc.wind} km/j
                            </span>
                          </div>
                        ))}
                      </div>
                    </div>
                  </>
                ) : (
                  <div className="flex items-center gap-2 text-destructive py-8 justify-center mx-auto">
                    <AlertCircle size={16} />
                    <p className="text-xs font-bold">Gagal memuat widget ramalan cuaca.</p>
                  </div>
                )}
              </CardContent>
            </Card>
          </motion.div>

        </div>
      </div>
    </>
  );
}
